### PR TITLE
Refs Task #387351 update user

### PR DESCRIPTION
**Pull request by nguyenthinhhn**
## Related Tickets

- [#387351](https://dev.sun-asterisk.com/issues/387351)

## What's this PR do ?

- [x] update user

## Performance

- [ ] Resolved n + 1 query
- [ ] Run explain query already
- [ ] Time run rake task : 1000 ms

## Checklist

- [x] It was tested in local success?
- [x] Fill link PR into ticket and the opposite
- [x] Note purpose/reason, solution, scope of influence into ticket

## ENV note

```
```

## Notes
Screenshot from 2019-12-06 15-14-54

### CODE REVIEW

**Comment by framgiaciv3** [comment]
No data coverage!

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr></table>

File: app/Http/Requests/UserUpdateRequest.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    
+    public function messages()
+    {
+        return [
+            'name.required' => 'Please enter name',
+            'name.min' => 'Name is too short (minimum is 5 characters)',
+            'name.max' => 'Name is too long (maximum is 50 characters)',
+            'email.required' => 'Please enter email',
+            'email.unique' => 'Email is unique',
```
Comment (tamtt-1031) [inline app/Http/Requests/UserUpdateRequest.php]: message này ko giống message màn hình khác nhỉ :v

Comment (nguyenthinhhn) [reply app/Http/Requests/UserUpdateRequest.php]: name thì giống còn email này thì giống vs màn hình create user, các màn hình khác có check email đâu a

File: app/Repositories/Eloquents/UserRepository.php [hunk 2]
```diff
@@ -50,4 +50,22 @@ private function handleSearchParams($searchParams)
 
         return $searchParams;
     }
+
+    public function updateUser($id, $request)
```
Comment (ghost) [inline app/Repositories/Eloquents/UserRepository.php]: đặt `update` là được r e

Comment (nguyenthinhhn) [reply app/Repositories/Eloquents/UserRepository.php]: ok a

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

File: app/Http/Requests/UserUpdateRequest.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return [
+            'name' => 'required|min:5|max:50|',
+            'email' => [
+                'required',
+                'max:200',
+                Rule::unique('users')->ignore($this->id),
+            ],
+            'password' => 'required|min:8|max:50|',
+            'avatar' => 'image',
```
Comment (ghost) [inline app/Http/Requests/UserUpdateRequest.php]: e add validte thêm đuôi .png, jpg, gif, jpeg luôn

File: public/js/user.js [hunk 4]
```diff
@@ -12,6 +12,20 @@ $('#avatar').change(function () {
...
+$('#avatar_edit').change(function () {
+    if ($(this).val() != '') {
+        var reader = new FileReader();
+        reader.onload = function (e){
+            $('.avatar_show_edit').attr('src', e.target.result);
+        }
+        reader.readAsDataURL(this.files[0]);
+    }
+})
+
```
Comment (ghost) [inline public/js/user.js:34]: trường hợp validate lỗi thì cái image user vừa chọn có bị mất không e

Comment (nguyenthinhhn) [reply public/js/user.js:34]: hiện  tại thì nó mất, để e fix

File: app/Http/Requests/UserUpdateRequest.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'name' => 'required|min:5|max:50|',
```
Comment (phucqd-1305) [inline app/Http/Requests/UserUpdateRequest.php]: thừa cái dấu | cuối cùng.

File: app/Http/Requests/UserUpdateRequest.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            'email' => [
+                'required',
+                'max:200',
+                Rule::unique('users')->ignore($this->id),
+            ],
+            'password' => 'required|min:8|max:50|',
```
Comment (phucqd-1305) [inline app/Http/Requests/UserUpdateRequest.php]: cái này nữa thừa |

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr></table>

File: app/Http/Requests/UserUpdateRequest.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    public function messages()
+    {
+        return [
+            'name.required' => 'Please enter name',
+            'name.min' => 'Name is too short (minimum is 5 characters)',
+            'name.max' => 'Name is too long (maximum is 50 characters)',
+            'email.required' => 'Please enter email',
+            'email.unique' => 'Email is unique',
```
Comment (tamtt-1031) [inline app/Http/Requests/UserUpdateRequest.php]: `This user email has already been used by another user`
ý a là dùng kiểu format thông báo này cho giống với những trường hợp unique trong hệ thống hiện tại

Comment (tamtt-1031) [reply app/Http/Requests/UserUpdateRequest.php]: a chưa thấy message này được định nghĩa trong spec. e bổ sung vào spec luôn nha

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr></table>

File: resources/views/admins/users/edit.blade.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+			<div class="col-xs-12">
+				<button class="btn btn-sm btn-warning float-right cancel-btn"><i class="fa fa-times"></i> Cancel</button>
+				<button type="submit" class="btn btn-sm btn-primary float-right"><i class="fa fa-check"></i> Save</button>
+			</div>
+			<div class="col-md-4">
+				<div class="form-group padding-bottom-30">
+					
+					@if($user->avatar)
+                            <img class="crusor width350 avatar_show_edit" src="/storage/{{ $user->avatar }}">
+                        @else
```
Comment (ghost) [inline resources/views/admins/users/edit.blade.php]: indent ko e

File: resources/views/admins/users/edit.blade.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+							<option value="{{ UserType::USER }}" {{ $user->role == UserType::USER ? "selected" : "" }}>User</option>
+							<option value="{{ UserType::ADMIN }}" {{ $user->role == UserType::ADMIN ? "selected" : "" }}>Admin</option>
+						</select>
+					</div>
+				</div>
+			</div>
+		</div>
+	</div>
+	{{ Form::close() }}
+    @include('modals.cancel_modal')
```
Comment (ghost) [inline resources/views/admins/users/edit.blade.php]: indent

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr></table>

File: public/js/user.js [hunk 10]
```diff
@@ -12,10 +18,55 @@ $('#avatar').change(function () {
...
+    $.ajax({
+        url: "/admin/users/" + id, 
+        data: formData,
+        type: 'POST',
+        contentType: false,
+        processData: false,
+        success: function (data) {
+            if(data.error == true) {
+                toastr.error(data.messageFail)
```
Comment (ghost) [inline public/js/user.js:57]: test case này chưa e

**Review by tamtt-1031 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: app/Http/Controllers/UserController.php [hunk 11]
```diff
@@ -4,8 +4,11 @@
 
 use Illuminate\Http\Request;
 use App\Http\Requests\UserCreateRequest;
+use App\Http\Requests\UserUpdateRequest;
 use App\Repositories\Interfaces\UserRepositoryInterface as UserRepository;
 use Illuminate\Support\Facades\Mail;
+use App\Models\User;
+use Session;
```
Comment (ghost) [inline app/Http/Controllers/UserController.php]: ko dùng thì remove luôn e

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr></table>

**Review by ghost — APPROVED** [review]

**Comment by phucqd-1305** [comment]
OK

**Pull request merged**